src/Interpreters/Session.h

```diff
@@ -0,0 +1,82 @@
+#pragma once
+
+#include <common/types.h>
+#include <Interpreters/Context_fwd.h>
+#include <Interpreters/ClientInfo.h>
+
+#include <chrono>
+#include <memory>
+#include <optional>
+
+namespace Poco::Net { class SocketAddress; }
+
+namespace DB
+{
+class SessionLog;
+class Credentials;
+struct NamedSession;
+class ContextAccess;
+struct Settings;
+class Authentication;
+
+/** Represents user-session from the server perspective,
+ *  basically it is just a smaller subset of Context API, simplifies Context management.
+ *
+ * Holds session context, facilitates acquisition of NamedSession and proper creation of query contexts.
+ * Adds log in, log out and login failure events to the SessionLog.
+ */
+class Session
+{
+public:
+//    static Session makeSessionFromCopyOfContext(const ContextPtr & _context_to_copy);
+    Session(const ContextPtr & context_to_copy, ClientInfo::Interface interface, std::optional<String> default_format = std::nullopt);
+    ~Session();
+
+    Session(const Session &) = delete;
+    Session& operator=(const Session &) = delete;
+
+    Session(Session &&);
+    Session& operator=(Session &&);
+
+    Authentication getUserAuthentication(const String & user_name) const;
```

Comment:
Why does the function `getUserAuthentication()` exist? I think we already have the function `setUser()` for checking password.

Reply:
There are some complex authentication scenarios, e.g. in PostgreSql or MySQL handlers:

https://github.com/ClickHouse/ClickHouse/blob/b29804752717749cb3f1056c2f2611219d92e58c/src/Server/MySQLHandler.cpp#L255